src/http/middleware/waf.rs

```diff
@@ -0,0 +1,1001 @@
+use std::{
+    net::IpAddr,
+    num::NonZeroU32,
+    str::FromStr,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use ahash::RandomState;
+use anyhow::{Context as _, anyhow};
+use arc_swap::ArcSwap;
+use async_trait::async_trait;
+use axum::{
+    extract::Request as AxumRequest,
+    response::{IntoResponse, Response as AxumResponse},
+};
+use futures::future::BoxFuture;
+use governor::{
+    Quota, RateLimiter,
+    clock::{Clock, DefaultClock},
+    state::{InMemoryState, NotKeyed, keyed::DashMapStateStore},
+};
+use http::{
+    HeaderMap, HeaderName, HeaderValue, Method, Request, Response, StatusCode, header::RETRY_AFTER,
+};
+use humantime::parse_duration;
+use itertools::Itertools;
+use regex::Regex;
+use serde::Deserialize;
+use serde_with::{DeserializeFromStr, DisplayFromStr, serde_as};
+use tokio_util::sync::CancellationToken;
+use tower::{Layer, Service};
+use url::Url;
+
+use crate::{
+    http::{Client, Error, middleware::extract_ip_from_request},
+    tasks::Run,
+};
+
+/// Matches HTTP status codes or ranges
+#[derive(Debug, Clone, Copy, Eq, PartialEq, DeserializeFromStr)]
+pub struct StatusRange {
+    from: u16,
+    to: Option<u16>,
+}
+
+impl StatusRange {
+    /// Check status code against the range
+    pub const fn evaluate(&self, v: StatusCode) -> bool {
+        let code = v.as_u16();
+
+        if let Some(to) = self.to {
+            return code >= self.from && code <= to;
+        }
+
+        code == self.from
+    }
+}
+
+impl FromStr for StatusRange {
+    type Err = crate::http::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```

Comment:
How about adding some check that from <= to?

Reply:
Makes sense!